### PR TITLE
tsdb: error on ingesting series with duplicate labels

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

xx

Alternative to #6660

### CODE REVIEW

File: pkg/labels/labels.go [hunk 1]
```diff
@@ -200,6 +200,18 @@ func (ls Labels) Has(name string) bool {
 	return false
 }
 
+// HasDuplicateLabels returns whether ls has duplicate label names.
+func (ls Labels) HasDuplicateLabelNames() bool {
```
Comment (roidelapluie) [inline pkg/labels/labels.go]: We can chose here if we want more CPU (iter twice on the labels)
or more RAM (found:=make([]string,0,len(ls)))

Comment (brian-brazil) [reply pkg/labels/labels.go]: Labels are sorted, so you can do this in O(n) time.

File: tsdb/head_test.go [hunk 2]
```diff
@@ -1290,3 +1290,21 @@ func TestNewWalSegmentOnTruncate(t *testing.T) {
...
+		testutil.Ok(t, os.RemoveAll(dir))
+	}()
+	wlog, err := wal.NewSize(nil, nil, dir, 32768, false)
+	testutil.Ok(t, err)
+
+	h, err := NewHead(nil, nil, wlog, 1000)
+	testutil.Ok(t, err)
+	defer h.Close()
+	app := h.Appender()
+	_, err = app.Add(labels.Labels{{Name: "a", Value: "c"}, {Name: "a", Value: "b"}}, 0, 0)
```
Comment (brian-brazil) [inline tsdb/head_test.go]: Test for a=b,a=b too

Comment (roidelapluie) [reply tsdb/head_test.go]: Should it error?

**Comment by mback2k** [comment]
> Alternative to #6660

Wouldn't it make prometheus more robust if path PRs were merged?

File: tsdb/head_test.go (hunk 2 above)
Comment (brian-brazil) [reply tsdb/head_test.go]: Yes

**Comment by roidelapluie** [comment]
> > Alternative to #6660
> 
> Wouldn't it make prometheus more robust if path PRs were merged?

No, but it would make it slower.

**Comment by roidelapluie** [comment]
I guess that @brian-brazil already knows it but after reading the code this looks like the good place to add this.

The Add() function will only be called once per metric per target. As metrics should not change a lot between scrapes, it will cause an overhead on the first scrape, but won't re-check for duplicate labels on next scrapes.

Adding that in the parser like in #6660 would mean that we make the work for every metric in every scrape.

**Comment by roidelapluie** [comment]
I guess that @brian-brazil already knows it but after reading the code this looks like the good place to add this.

The Add() function will only be called once per metric per target. As metrics should not change a lot between scrapes, it will cause an overhead on the first scrape, but won't re-check for duplicate labels on next scrapes.

Adding that in the parser like in #6660 would mean that we make the work for every metric in every scrape.

File: tsdb/head.go [hunk 3]
```diff
@@ -921,6 +925,10 @@ func (a *headAppender) Add(lset labels.Labels, t int64, v float64) (uint64, erro
 	// Ensure no empty labels have gotten through.
 	lset = lset.WithoutEmpty()
 
+	if l, dup := lset.HasDuplicateLabelNames(); dup {
+		return 0, errors.Wrap(ErrInvalidSample, fmt.Sprintf(`label "%s" is not unique`, l))
```
Comment (brian-brazil) [inline tsdb/head.go]: label name

**Comment by brian-brazil** [comment]
We should add a test to confirm that this causes such a scrape to fail.

**Comment by roidelapluie** [comment]
Test added.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**